### ISSUE TITLE
"Save best" and other improvements to model checkpointing

### ISSUE THREAD

**Issue by fynnsu** [vllm-project/speculators#307]
Our current system for saving checkpoints is pretty basic and while it works it could be improved. Currently, we just save (in the checkpoint folder) a new folder for each epoch. 

E.g.
```
../path/
    0/          # epoch number
        config.json
        model.safetensors
        optimizer_state_dict.pt
        scheduler_state_dict.pt (optional)
    1/
        config.json
        model.safetensors
        optimizer_state_dict.pt
        scheduler_state_dict.pt (optional)
    ...
```

Then if continuing training we just look for folders with numbered names in the save dir and load the highest numbered one.

It would be good if we could improve this system. One thing that comes to mind is having a "save_best" option where we only save an epoch if it has the lowest validation loss. 

Additionally, the optimizer state dict and scheduler state dict are useful if resuming training, but not everyone is interested in doing so and they may need to be manually removed before transferring the checkpoint (e.g. uploading to huggingface). It might be better to only optionally save them when the user requests it.

**Comment by fynnsu** [comment]
Hi @surojitiitg, thanks for taking up the issue! And yes, I intended this issue to be a little bit open ended, so if you have suggestions I'm happy to hear them!

**Comment by fynnsu** [comment]
@NJX-njx, this is already being worked on by @surojitiitg. Please don't work on issues that have already been assigned.
